### PR TITLE
Implement ATen Distributions.cu + Poisson

### PR DESCRIPTION
**Pull request by rachtsingh**
I got a bit into the weeds and the scope expanded, but I think this is the right way to do it. This PR (will) implement a pointwise Poisson sampling method for CPU/CUDA, and should be followed up quickly with ports of the Gamma / Dirichlet samplers, and other things we might want to use from `distributions.c`.

So far:
- [x] Renames `Generator` inside `THCRandomTensor.h` to be `THCGenerator`, and fix CUDAGenerator instantiation
- [x] Implement Poisson distribution + `torch.poisson` (Variable only for now, but we can wrap it). 
- [x] Write tests for the distribution.
- [x] Implement a Philox-based sampler for rejection sampling.

### CODE REVIEW

**Comment by rachtsingh** [comment]
Ah, not quite ready for review yet, by the way - I still need to debug this CUDA issue and write tests.

File: aten/src/THC/THCTensorRandom.cu [hunk 1]
```diff
@@ -97,34 +175,48 @@ __device__ inline half half_uniform_scale_and_shift(float x, double a, double b)
...
-  int idx = blockIdx.x * BLOCK_SIZE + threadIdx.x;                             \
-  int rounded_size = THCCeilDiv(size, BLOCK_SIZE) * BLOCK_SIZE;                \
-  for (int i = idx; i < rounded_size; i += BLOCK_SIZE * MAX_NUM_BLOCKS) {      \
-    CURAND_T x = CURAND_FUNC(&state[blockIdx.x]);                              \
-    if (i < size) {                                                            \
-      T y = TRANSFORM;                                                         \
-      result[i] = y;                                                           \
-    }                                                                          \
-  }                                                                            \
+#define GENERATE_KERNEL1(NAME, T, ARG1, CURAND_T, CURAND_FUNC, TRANSFORM)           \
```
Comment (fritzo) [inline aten/src/THC/THCTensorRandom.cu]: nit: Consider reverting unnecessary whitespace changes to ease review by PyTorch folks.

Comment (rachtsingh) [reply aten/src/THC/THCTensorRandom.cu]: Thanks! I usually squash before review but I'll be more on top of it.

**Comment by gchanan** [comment]
On your question, I don't think we should add calls to get random numbers directly to the generator; the normal pattern is to pass in the generator to a function that generates the random numbers, right?

**Comment by rachtsingh** [comment]
Yeah, that sounds right. I'm not sure about casting the result of `unsafeGetTH` from `void *` to be able to sample though - is it ok to change the generator to output the specialized `Generator` when parsing `native_functions.yaml`?

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh [hunk 2]
```diff
@@ -283,12 +283,14 @@ bool CUDA_tensor_apply2(at::Tensor a,
...
-    return false;
-  }
+  const dim3 block = dim3(256);
+  // const dim3 block = getApplyBlock();
+
+  dim3 grid = dim3(64);
+  //g if (!getApplyGrid(totalElements, grid)) {
+  //g   return false;
+  //g }
+  //g grid = dim3(1);
```
Comment (rachtsingh) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: This is temporary - looking for advice on how to best limit the blockDim to 256 when calling `CUDA_tensor_apply2 `

**Comment by rachtsingh** [comment]
Ok, I think it's probably ready for review. @fritzo, given the C++/CUDA changes, it's probably best to  ask someone from PyTorch to review as well? I don't know what exactly the plans are for CUDAGenerator, etc.

File: torch/distributions/poisson.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+def _poisson(_lambda):
+    if not isinstance(_lambda, Variable):
+        return torch.poisson(Variable(_lambda)).data
+    return torch.poisson(_lambda)
+
+
+class Poisson(Distribution):
+    r"""
+    Creates a Poisson distribution parameterized by `_lambda`, the rate parameter.
```
Comment (fritzo) [inline torch/distributions/poisson.py]: nit: preceding underscore usually denotes a private variable. It's probably safer to call this `lam` or `lambda_`. Alternatively it would be nice to call it `rate` which is both semantically meaningful and improves compatibility with tensorflow distributions.

Comment (fritzo) [reply torch/distributions/poisson.py]: For example `self._lambda` would be the obvious member name but that looks private. Instead `self.lambda_` or `self.rate` are clearly public.

**Comment by fritzo** [comment]
@apaszke How should we test the CUDA sampler? Do we use the same tests in `test_distributions.py` that we use for CPU samplers, or should these new tests live in `test_cuda.py`?

File: torch/distributions/poisson.py (hunk 3 above)
Comment (rachtsingh) [reply torch/distributions/poisson.py]: Both sound good to me. Let's use `rate` since that's more evocative? I.e. like `mean`/`std` in `Normal` already, and matches up with `Exponential`, which is the right idea.

**Comment by apaszke** [comment]
I think `test_distributions.py` is ok. Just make `unittest` skip the tests if CUDA is unavailable (look at `test_nn.py` and `test_autograd.py` for examples).

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+          ret_val = curand_poisson(&states[blockIdx.x], lambda);
+        }
+      );
+    }
+  };
+
+} // at::native::dist
+
+Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
+  Tensor ret = lambda.type().toScalarType(kDouble).zeros(lambda.sizes());
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Distributions.cu]: It's better to use `tensor` instead of `zeros` if you don't depend on the initial values of the tensor (it will be uninitialized)

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  auto lambda_ = lambda.toType(ScalarType::Double);
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Distributions.cu]: Is this unstable in fp32? fp64 math is extremely slow on non-Tesla GPUs, and we should avoid it

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Distributions.cu]: Is this the right place for this copyright notice?

File: torch/csrc/Module.cpp [hunk 7]
```diff
@@ -750,6 +751,7 @@ static PyMethodDef TorchMethods[] = {
   {"_standard_gamma",  (PyCFunction)THPModule_standard_gamma,    METH_VARARGS | METH_KEYWORDS, NULL},
   {"_dirichlet_grad", (PyCFunction)THPModule_dirichlet_grad,    METH_VARARGS | METH_KEYWORDS, NULL},
   {"bernoulli",       (PyCFunction)THPModule_bernoulli,         METH_VARARGS | METH_KEYWORDS, NULL},
+  {"poisson",         (PyCFunction)THPModule_poisson,           METH_VARARGS | METH_KEYWORDS, NULL},
```
Comment (apaszke) [inline torch/csrc/Module.cpp]: Let's avoid adding more of the sampling methods to the global namespace. `torch.distributions` is the way to expose those samplers

**Comment by rachtsingh** [comment]
Ok, think I addressed comments (thanks for the review!) except for Adam's comment about cluttering the global namespace. What's the right way to do that? I think everything from `._C` gets imported automatically, so I'm not sure which config file to edit to expose it to Python.

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 8]
```diff
@@ -40,12 +31,13 @@ namespace dist {
...
     }
   };
+
 } // at::native::dist
 
 Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
   Tensor ret = lambda.type().tensor(lambda.sizes());
   auto lambda_ = lambda.toType(ScalarType::Float);
-  dispatch_all<void, dist::PoissonOpCUDA>(ret.type(), "poisson", ret, lambda_, dist::get_states(gen));
+  dispatch_floating_types<void, dist::PoissonOpCUDA>(ret.type(), "poisson", ret, lambda_, dist::get_states(gen));
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/Distributions.cu]: if you are having trouble with half, try `DISPATCH_ALL_FLOATING_TYPES`; `dispatch_all` uses `Half` and `DISPATCH_ALL_FLOATING_TYPES` uses `half`; we should resolve this, probably always using `Half` (and providing __device__ conversion functions?).  CC @colesbury

Comment (rachtsingh) [reply aten/src/ATen/native/cuda/Distributions.cu]: Thanks, that works!

Comment (apaszke) [reply aten/src/ATen/native/cuda/Distributions.cu]: Ideally we would always use `half` for device functions and `at::Half` for host functions.

**Comment by rachtsingh** [comment]
After doing some checks, it looks like `curand_poisson` without using the more robust API uses either an inverse CDF sampler or the transformed rejection method, neither of which are accurate for low values of `lambda`. So, it fails the distribution tests (good thing they work!). 

I think there's 3 options - (1) to merge as-is (by raising the threshold on the CUDA version), but to warn that for low values of `lambda`, the results may be biased, or (2) merge the CPU version, not the GPU version, or (3) punt on this until we get Philox-based PRNGs up, which was what I was planning to tackle next.

**Comment by apaszke** [comment]
I'm fine with all options. I'll leave the decision to @fritzo

**Comment by fritzo** [comment]
Option 1 sounds very reasonable. You could also [file an issue at probtorch/pytorch](https://github.com/probtorch/pytorch/issues/new) noting that we should revisit this.

**Comment by rachtsingh** [comment]
OK, fixed the issues and added a Philox-based RNG. I don't think the sampling algorithm is the most accurate, but we can work on it in the future I think.

Also, I think there's a bug in the handling of `failure_rate` (note that I *decreased* that to make the tests pass), but I'll have to double check the math before fixing.

**Comment by rachtsingh** [comment]
(discussing in https://github.com/pytorch/pytorch/pull/4556)
